### PR TITLE
Advanced Provider Support

### PR DESCRIPTION
**Pull request by gokmen**
In this PR we re-shape whole provider support on client side and backend part. And introducing support for `Google Compute Engine` and `DigitalOcean` providers.

full screencast, wip: http://recordit.co/5WPY15PLfg

image
## Todo
- [x] Onboarding wizard needs to be simplified
- [x] Application should rely one place for all provider settings
- [x] Stack-Editor should rely on templates provided per provider
- [x] Template/Provider structure should be revised
- [x] Google Provider template needs to be provided
- [x] Azure Provider template needs to be provided
- [x] DigitalOcean Provider template needs to be provided
- [x] Existing providers (AWS, Vagrant) should be updated for new structure
- [x] Actions per provider (create, modify, setAlwaysOn etc.) needs to be revised
  - [x] AWS Provider
  - [x] Vagrant Provider
  - [x] Google Provider
  - [x] DigitalOcean Provider
  - [x] Azure Provider
- [x] Provider parser should be unified and support all provider types
- [x] Stack build modal should be unified and support all supported providers
- [x] Stack-Editor should warn user if there is a change happened on current stack template cc/ @usirin 
- [x] Credential views should use the data provided in provider config
- [x] Routing issue with Dashboard and Stack-Editor needs to be fixed
## Types of changes
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:
- [x] My change requires a change to the documentation.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by gokmen** [comment]
This is currently not `wip` it's waiting for following PR's to be merged and tested with this branch. @sinan you can review it as is, but do not merge it yet even if it's ready.

https://github.com/koding/koding/pull/9324 , https://github.com/koding/koding/pull/9331 are required for `GCE` and also if we take https://github.com/koding/koding/pull/9241 in we can include `Azure Provider` support into this one as well.

File: client/app/lib/providers/resourcestatemodal/views/baseerrorpageview.coffee [hunk 1]
```diff
@@ -33,18 +33,26 @@ module.exports = class BaseErrorPageView extends JView
...
+      @errorContainer.wrapper.scrollToBottom()
 
-  onPageDidShow: ->
 
+  onPageDidShow: -> kd.utils.defer ->
     # it needs to update container height if it can't be set fixed in css.
     # otherwise, custom scroll doesn't work properly
     container = @getDomElement().find '.main'
-    container.css 'height', container.height()
+    container.css 'height', container.height() or 400
```
Comment (sinan) [inline client/app/lib/providers/resourcestatemodal/views/baseerrorpageview.coffee]: this doesn't seem to fix the root issue here, maybe we can create a task for this?

File: client/stack-editor/lib/index.coffee [hunk 2]
```diff
@@ -166,34 +164,37 @@ module.exports = class StackEditorAppController extends AppController
...
+              @openEditor stackTemplate._id
+          cancel:
+            title: 'Cancel'
+            style: 'solid light-gray medium'
+            callback: ->
+              confirmation.destroy()
+              confirmation = null
+        content: '''
+          <div class='modalformline'
+            <p>This stack template is updated by another admin. Do you want to reload?
```
Comment (sinan) [inline client/stack-editor/lib/index.coffee]: `</p>`

File: client/stack-editor/lib/styl/stackwizard.styl [hunk 3]
```diff
@@ -183,70 +177,59 @@ $accentColor                = #F2778A
...
+    &.azure
+      r-sprite              'stack-editor' 'azure'
+    &.rackspace
+      r-sprite              'stack-editor' 'rackspace'
+    &.softlayer
+      r-sprite              'stack-editor' 'softlayer'
+
+  .beta:before
+  .coming-soon:before
+    content                 'BETA'
```
Comment (sinan) [inline client/stack-editor/lib/styl/stackwizard.styl]: for the localization purposes you can use html data attributes in `content` such as `content: attr(data-beta);`

Comment (sinan) [inline client/stack-editor/lib/styl/stackwizard.styl]: or even better to generalize:

```html
<p data-before-content='BETA'></p>
```

```css
content: attr(data-before-content)
```

File: config/credentials.default.coffee [hunk 4]
```diff
@@ -180,7 +180,7 @@ module.exports = (options) ->
   dummyAdmins = ['superadmin', 'admin', 'koding']
   druid =
     host : options.serviceHost
-    port : 8090
+    port : '8090'
```
Comment (sinan) [inline config/credentials.default.coffee]: 👍

**Review by sinan — COMMENTED** [review]
Looks good to me! 💯 just minor stuff

File: client/app/lib/providers/resourcestatemodal/views/baseerrorpageview.coffee (hunk 1 above)
Comment (gokmen) [inline client/app/lib/providers/resourcestatemodal/views/baseerrorpageview.coffee]: yes, `defer` in here solves the issue https://github.com/koding/koding/pull/9319#discussion-diff-84153652R51 but in any case I added that one as well if I missed something.

**Comment by gokmen** [comment]
@sinan all issues are fixed 👍

**Comment by gokmen** [comment]
@sinan can u review one more time, not much changed in structure just added more providers. this is ready to go but it's depending on #9453 so better to wait that one before merge. I already tested this branch with changes in #9453

File: client/app/lib/providers/templates/aws.coffee [hunk 5]
```diff
@@ -33,7 +38,7 @@ provider:
 resource:
   aws_instance:
     # this is the name of your VM
-    example-instance:
+    aws-instance:
```
Comment (sinan) [inline client/app/lib/providers/templates/aws.coffee]: can we use a dictionary here, like you did with the stack names?

File: client/app/lib/providers/templates/azure.coffee [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+provider:
+  azure:
+    publish_settings: '${var.azure_publish_settings}'
+    subscription_id: '${var.azure_subscription_id}'
+
+resource:
+  azure_instance:
+    # this is the name of your VM
+    azure-instance:
```
Comment (sinan) [inline client/app/lib/providers/templates/azure.coffee]: again maybe dictionary.

File: workers/social/lib/social/models/computeproviders/credential.coffee [hunk 7]
```diff
@@ -485,7 +485,7 @@ module.exports = class JCredential extends jraphical.Module
 
   # Poor man's shadow function ~ GG
```
Comment (sinan) [inline workers/social/lib/social/models/computeproviders/credential.coffee]: :)

**Review by sinan — APPROVED** [review]
👍

File: client/app/lib/providers/templates/aws.coffee (hunk 5 above)
Comment (gokmen) [inline client/app/lib/providers/templates/aws.coffee]: sure, I'll make it happen in next release 👍

**Pull request merged**